src/main/java/htsjdk/samtools/cram/structure/Slice.java

```diff
@@ -318,4 +333,113 @@ public CRAIEntry getCRAIEntry() {
     public CRAIEntry getCRAIEntry(final long containerStartOffset) {
         return new CRAIEntry(sequenceId, alignmentStart, alignmentSpan, containerStartOffset, offset, size);
     }
+
+    /**
+     * Create a Slice from CRAM Compression Records and a Compression Header
+     *
+     * @param records input CRAM Compression Records
+     * @param header the enclosing {@link Container}'s Compression Header
+     * @return a Slice corresponding to the given records
+     */
+    public static Slice buildSlice(final List<CramCompressionRecord> records,
+                                   final CompressionHeader header) {
+        final Map<Integer, ByteArrayOutputStream> externalBlockMap = new HashMap<>();
+        for (final int id : header.externalIds) {
+            externalBlockMap.put(id, new ByteArrayOutputStream());
+        }
+
+        final Slice slice = new Slice();
+        slice.nofRecords = records.size();
+
+        int minAlStart = Integer.MAX_VALUE;
+        int maxAlEnd = SAMRecord.NO_ALIGNMENT_START;
+        {
+            // @formatter:off
+            /*
+             * 1) Count slice bases.
+             * 2) Decide if the slice is single ref, unmapped or multi reference.
+             * 3) Detect alignment boundaries for the slice if not multi reference.
+             */
+            // @formatter:on
+
+            // Valid Slice states, by individual record contents:
+            //
+            // Single Reference: all records have valid placements/alignments on the same reference sequence
+            //      * records can be unmapped-but-placed
+            //      * reference can be external or embedded
+            // Multiple Reference: records may be placed or not, and may have differing reference sequences
+            //      * reference must not be embedded (not checked here)
+            // Unmapped: all records are unmapped and unplaced
```

Comment:
This is the definition of the "slice state" enum and state variable I was asking about above. Should there just be an enum for these values, and a single state variable that changes as needed ? And also one for container, perhaps with slightly different states. That would codify the whole state thing nicely.